src/weekly/query.js

```diff
@@ -58,13 +58,15 @@ async function queryZuora (deliveryDate, config: Config) {
            ) OR
           (
             RatePlan.AmendmentType = 'RemoveProduct' AND 
-            Amendment.EffectiveDate > '${formattedDeliveryDate}' 
+            Amendment.EffectiveDate > '${formattedDeliveryDate}' AND
+            EffectiveStartDate <= '${formattedDeliveryDate}' 
             ) OR 
            (
             RatePlan.AmendmentType = 'NewProduct' AND 
             Amendment.CustomerAcceptanceDate <= '${formattedDeliveryDate}' 
            )
         ) AND
+        IsLastSegment = true AND
```

Comment:
it looks like IsLastSegment checks the EffectiveStartDate and EffectiveEndDate to make sure it's current, rather than being the last one overall, is that your understanding?
https://knowledgecenter.zuora.com/Central_Platform/API/G_SOAP_API/E1_SOAP_API_Object_Reference/RatePlanCharge
if we are now looking at the charge dates directly, do we need to look at the amendments at all?

Reply:
Based on [charge-detail docs](https://www.zuora.com/developer/api-reference/#operation/GET_SubscriptionsByKey`) I think `IsLastSegment` means the last **version** of the segment, not the **current** segment active today:

> The segmented rate plan charges. When an amendment results in a change to a charge, Zuora creates a segmented rate plan charge. Use this field to track segment charges.
> 
> Possible values are:
> 
> **last-segment**: (Default) The last rate plan charge on the subscription. The last rate plan charge is the last one in the order of time on the subscription rather than the most recent changed charge on the subscription.
>
> **current-segment**: The segmented charge that is active on today’s date (effectiveStartDate <= today’s date < effectiveEndDate).
> 
> **all-segments**: All the segmented charges. The chargeSegments field is returned in the response. The chargeSegments field contains an array of the charge information for all the charge segments.
> 
> **specific-segment&as-of-date=date**: The segmented charge that is active on a date you specified ((specific date = effectiveStartDate) OR (effectiveStartDate < specific date < effectiveEndDate)). The format of the date is yyyy-mm-dd.

Now, neither last segment, nor current segment, are actually 100% correct, instead we would need something like

```
AND (EffectiveStartDate >= '${formattedDeliveryDate}' AND EffectiveEndDate <= '${formattedDeliveryDate}')
```

however because `formattedDeliveryDate` is one week in the future from today, then there might not be yet such a RatePlanCharge (if we are less than week away from `EffectiveEndDate`). Addressing this would likely mean much bigger change to the existing query.

So if `IsLastSegment` is not strictly correct, then why use it? Well I have compared RatePlan-based-before extract and RatePlanCharge-with-IsLastSegment-after extract and the diff of >46K records was identical. So at least there should be no further bugs introduces (although we might be carrying over existing ones :)

> if we are now looking at the charge dates directly, do we need to look at the AmendmentType

Indeed it is likely we can clean up the query, however as it is risky to change these queries fearing breaking fulfilment I have tried to make the smallest change possible in the current PR. Hopefully if regression is introduced, then it would be just in relatively smaller subset of records previously selected by sub-clause

```
RatePlan.AmendmentType = 'RemoveProduct' AND 
Amendment.EffectiveDate > '${formattedDeliveryDate}'
```

Maybe we could try improving the query further in the (mythical) second iteration.

Reply:
I think that's a very helpful explanation and well done for finding the documentation.  I agree with your motivation to make the smallest possible change.
When this was written, Patricio spent months doing almost exactly the process you are doing of diffing the old and new fulfilment files until all the differences were due to the old one being incorrect.  So you are following exactly the same process which has served well. 👍